src/librustc_mir/interpret/intrinsics.rs

```diff
@@ -316,9 +316,44 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
                 self.write_scalar(offset_ptr, dest)?;
             }
             sym::ptr_guaranteed_eq | sym::ptr_guaranteed_ne => {
-                // FIXME: return `true` for at least some comparisons where we can reliably
-                // determine the result of runtime (in)equality tests at compile-time.
-                self.write_scalar(Scalar::from_bool(false), dest)?;
+                let a = self.read_immediate(args[0])?.to_scalar()?;
+                let b = self.read_immediate(args[1])?.to_scalar()?;
+
+                let geq_mode = intrinsic_name == sym::ptr_guaranteed_eq;
+                // `Some` in `cmp` means the comparison is known, guaranteed in both directions
+                let cmp = match (a, b) {
+                    // Comparisons between integers are always known.
+                    (Scalar::Raw { .. }, Scalar::Raw { .. }) => {
+                        if geq_mode {
+                            a == b
+                        } else {
+                            a != b
+                        }
+                    }
+                    // Comparisons of abstarct pointers with null pointers are known if the pointer
+                    // is in bounds.
+                    (Scalar::Raw { data: 0, .. }, Scalar::Ptr(ptr))
+                    | (Scalar::Ptr(ptr), Scalar::Raw { data: 0, .. }) => {
+                        if geq_mode {
+                            // We cannot know whether a pointer is null.
+                            false
+                        } else {
+                            // If the abstract pointer is in bounds, it cannot be a null pointer.
+                            !self.memory.ptr_may_be_null(ptr)
```

Comment:
I think this code is not even tested currently? `is_null` calls the "eq" intrinsic. In fact `is_null` always returns "false" during CTFE currently when the input is a pointer.

Reply:
I added some tests now